crates/proto/src/xfer/dnssec_dns_handle.rs

```diff
@@ -1003,6 +1017,260 @@ pub fn verify_nsec(query: &Query, soa_name: &Name, nsecs: &[&Record]) -> Proof {
+    //
+    // 1. There's only one NSEC3 record matching the QNAME, valid for NO_DATA
+    //    responses
+    // 2. There's **1 or 2** records present aka "Closest Encloser Proof"
+    // 3. In addition to "Closest Encloser Proof" records pair there's a third
+    //    record for an encopassing wildcard
+
+    if nsec3s.len() == 1 {
+        let nsec3: &NSEC3 = nsec3s[0];
+        let hashed_qname = hash(query.name(), salt, iterations);
+        if hashed_qname == nsec3.next_hashed_owner_name() {
+            // TODO: It's valid for NO_DATA responses only
+            return Proof::Secure;
+        }
+    }
+
+    // Search for Closest Encloser Proof
+
+    // Situation 1: `x.exmaple.com` doesn't exist but `exmaple.com` does
+    // Situation 2: `x.exmaple.com` doesn't exist but `exmaple.com` and
+    //              `*.example.com` do
+    // Situation 3: `x.a.b.c.exmaple.com` doesn't exist but `c.exmaple.com` does
+    // Situation 4: `x.a.b.c.exmaple.com` doesn't exist but both
+    //              `c.exmaple.com` and `*.c.exmaple.com` do
+
+    let next_closer_candidate = loop {
+        let next_closer = query.name().base_name();
+        if next_closer == Name::root() {
+            // no enclosing names?
+            return Proof::Bogus;
+        }
+        let hashed_next_closer = hash(&next_closer, salt, iterations);
+        let record = nsec3s
+            .iter()
+            .find(|n| n.next_hashed_owner_name() == hashed_next_closer);
+        if record.is_some() {
+            break next_closer;
+        }
+    };
+    let closest_encloser_candidate = next_closer_candidate.base_name();
+    let hashed_closest_encloser_candidate = hash(&closest_encloser_candidate, salt, iterations);
+    let record = nsec3s
+        .iter()
+        .find(|n| n.next_hashed_owner_name() == hashed_closest_encloser_candidate);
+
+    // is it a two-record proof or a single-record proof?
+    let (closest_encloser, next_closer) = if record.is_some() {
+        (closest_encloser_candidate, Some(next_closer_candidate))
+    } else if query.name().base_name() == closest_encloser_candidate {
+        // The single-record proof means for `x.a.example.com` `a.example.com`
+        // exists. Having an NSEC3 for `example.com` and not `a.example.com`
+        // would be an error
+        (next_closer_candidate, None)
+    } else {
+        return Proof::Bogus;
+    };
+
+    // There may be a wildcard for a closest encloser proof
+    let wildcard_encloser = Name::new()
+        .append_label("*")
+        .unwrap()
+        .append_domain(&closest_encloser)
+        .unwrap();
+    let hashed_wildcard_encloser = hash(&wildcard_encloser, salt, iterations);
+
+    let wildcard_record = nsec3s
+        .iter()
+        .find(|n| n.next_hashed_owner_name() == hashed_wildcard_encloser);
+
+    // Check for number of records matching the expectation based on proof
+    // type:
+
+    let expected_record_count = match (next_closer, wildcard_record) {
```

Comment:
Could you add a reference or a description on what the combinations & their values mean?

Reply:
I've added some extra explanation above (lines 1204 and further) at the time of writing.